File: apps/backend/src/routes/api/v1/chats.ts
import { Hono } from 'hono';
import { zValidator } from '@hono/zod-validator';
import { chatService } from '../../../services/chat.service';
import { authMiddleware } from '../../../middleware/auth';
import {
  createChatSchema,
  updateChatSchema,
  addMessageSchema,
  updateMessageSchema,
} from '../../../utils/validators-extended';

const chatRouter = new Hono();

chatRouter.post('/', authMiddleware, zValidator('json', createChatSchema), async (c) => {
  const user = c.get('user');
  const { title } = c.req.valid('json');

  const chat = await chatService.createChat(user.id, title);
  return c.json(chat);
});

chatRouter.get('/', authMiddleware, async (c) => {
  const user = c.get('user');
  const folderId = c.req.query('folderId');
  const archived = c.req.query('archived');
  const pinned = c.req.query('pinned');

  const filters: any = {};
  if (folderId !== undefined) filters.folderId = folderId || null;
  if (archived !== undefined) filters.archived = archived === 'true';
  if (pinned !== undefined) filters.pinned = pinned === 'true';

  const chats = await chatService.getChats(user.id, filters);
  return c.json(chats);
});

chatRouter.get('/shared/:shareId', async (c) => {
  const shareId = c.req.param('shareId');
  const chat = await chatService.getChatByShareId(shareId);
  return c.json(chat);
});

chatRouter.get('/:id', authMiddleware, async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');

  const chat = await chatService.getChatById(id, user.id);
  return c.json(chat);
});

chatRouter.put('/:id', authMiddleware, zValidator('json', updateChatSchema), async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');
  const data = c.req.valid('json');

  const chat = await chatService.updateChat(id, user.id, data);
  return c.json(chat);
});

chatRouter.delete('/:id', authMiddleware, async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');

  await chatService.deleteChat(id, user.id);
  return c.json({ message: 'Chat deleted' });
});

chatRouter.post('/:id/messages', authMiddleware, zValidator('json', addMessageSchema), async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');
  const messageData = c.req.valid('json');

  const message = await chatService.addMessage(id, user.id, messageData);
  return c.json(message);
});

chatRouter.put('/:id/messages/:messageId', authMiddleware, zValidator('json', updateMessageSchema), async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');
  const messageId = c.req.param('messageId');
  const { content } = c.req.valid('json');

  const message = await chatService.updateMessage(id, user.id, messageId, content);
  return c.json(message);
});

chatRouter.put('/:id/archive', authMiddleware, async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');
  const body = await c.req.json();
  const archived = body.archived ?? true;

  const chat = await chatService.archiveChat(id, user.id, archived);
  return c.json(chat);
});

chatRouter.put('/:id/pin', authMiddleware, async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');
  const body = await c.req.json();
  const pinned = body.pinned ?? true;

  const chat = await chatService.pinChat(id, user.id, pinned);
  return c.json(chat);
});

chatRouter.post('/:id/share', authMiddleware, async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');

  const result = await chatService.shareChat(id, user.id);
  return c.json(result);
});

chatRouter.delete('/:id/share', authMiddleware, async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');

  await chatService.unshareChat(id, user.id);
  return c.json({ message: 'Chat unshared' });
});

export default chatRouter;

File: apps/backend-python/src/routes/api/v1/chats.py
from fastapi import APIRouter, Depends
from pydantic import BaseModel
from typing import Optional
from sqlalchemy.orm import Session
import httpx
from src.config.database import get_db
from src.middleware.auth import get_current_user
from src.services.chat_service import chat_service
from src.services.proxy_service import (
    _get_connection_by_model,
    _get_default_connection,
    _build_headers,
    _chat_url,
)
from src.db.models import User

router = APIRouter()


class CreateChatSchema(BaseModel):
    title: str


class UpdateChatSchema(BaseModel):
    title: Optional[str] = None
    chat: Optional[dict] = None
    folderId: Optional[str] = None
    meta: Optional[dict] = None


class AddMessageSchema(BaseModel):
    role: str
    content: str
    model: Optional[str] = None
    images: Optional[list] = None
    sources: Optional[list] = None


class UpdateMessageSchema(BaseModel):
    content: str


class ArchiveSchema(BaseModel):
    archived: bool = True


class PinSchema(BaseModel):
    pinned: bool = True


class GenerateTitleSchema(BaseModel):
    model: str
    message: str


@router.post("")
def create_chat(data: CreateChatSchema, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return chat_service.create_chat(db, user.id, data.title)


@router.get("")
def list_chats(
    folderId: Optional[str] = None,
    archived: Optional[str] = None,
    pinned: Optional[str] = None,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    filters = {}
    if folderId is not None:
        filters["folderId"] = folderId or None
    if archived is not None:
        filters["archived"] = archived == "true"
    if pinned is not None:
        filters["pinned"] = pinned == "true"
    return chat_service.get_chats(db, user.id, filters)


@router.get("/shared/{share_id}")
def get_shared_chat(share_id: str, db: Session = Depends(get_db)):
    return chat_service.get_chat_by_share_id(db, share_id)


@router.get("/{chat_id}")
def get_chat(chat_id: str, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return chat_service.get_chat_by_id(db, chat_id, user.id)


@router.put("/{chat_id}")
def update_chat(chat_id: str, data: UpdateChatSchema, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return chat_service.update_chat(db, chat_id, user.id, data.model_dump(exclude_none=True))


@router.delete("/{chat_id}")
def delete_chat(chat_id: str, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    chat_service.delete_chat(db, chat_id, user.id)
    return {"message": "Chat deleted"}


@router.post("/{chat_id}/messages")
def add_message(chat_id: str, data: AddMessageSchema, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return chat_service.add_message(db, chat_id, user.id, data.model_dump(exclude_none=True))


@router.put("/{chat_id}/messages/{message_id}")
def update_message(chat_id: str, message_id: str, data: UpdateMessageSchema, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return chat_service.update_message(db, chat_id, user.id, message_id, data.content)


@router.put("/{chat_id}/archive")
def archive_chat(chat_id: str, data: ArchiveSchema, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return chat_service.archive_chat(db, chat_id, user.id, data.archived)


@router.put("/{chat_id}/pin")
def pin_chat(chat_id: str, data: PinSchema, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return chat_service.pin_chat(db, chat_id, user.id, data.pinned)


@router.post("/{chat_id}/share")
def share_chat(chat_id: str, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return chat_service.share_chat(db, chat_id, user.id)


@router.delete("/{chat_id}/share")
def unshare_chat(chat_id: str, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    chat_service.unshare_chat(db, chat_id, user.id)
    return {"message": "Chat unshared"}


@router.post("/{chat_id}/title/generate")
def generate_chat_title(
    chat_id: str,
    data: GenerateTitleSchema,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Call AI to generate a short title for the chat, update and return it."""
    try:
        conn = (
            _get_connection_by_model(db, user.id, data.model)
            if data.model
            else _get_default_connection(db, user.id)
        )
        resp = httpx.post(
            _chat_url(conn),
            json={
                "model": data.model,
                "messages": [
                    {
                        "role": "system",
                        "content": (
                            "You are a title generator. Your only job is to create a short chat session title "
                            "based on the topic of the user's message below. "
                            "Rules: maximum 5 words, start with one relevant emoji, no quotes, no punctuation at the end, "
                            "do NOT answer the user's question, do NOT introduce yourself, "
                            "respond with ONLY the title text and nothing else."
                        ),
                    },
                    {
                        "role": "user",
                        "content": f"Generate a title for a conversation that starts with this message: {data.message[:500]}",
                    },
                ],
                "stream": False,
                "max_tokens": 20,
            },
            headers=_build_headers(conn),
            timeout=30.0,
        )
        resp.raise_for_status()
        title = (
            resp.json()
            .get("choices", [{}])[0]
            .get("message", {})
            .get("content", "")
            .strip()
        )
    except Exception:
        title = data.message[:50].strip()

    if title:
        chat_service.update_chat(db, chat_id, user.id, {"title": title})
    return {"title": title}
